File: app/src/main/java/com/example/coursecomrade/CourseDetailActivity.java
package com.example.coursecomrade;

import android.content.Intent;
import android.os.Bundle;
import android.util.Log;
import android.view.MenuItem;
import android.view.View;
import android.widget.ImageButton;
import android.widget.PopupMenu;
import android.widget.TextView;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.Toolbar;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class CourseDetailActivity extends AppCompatActivity {

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_course_detail);

        // Set up toolbar
        Toolbar toolbar = findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);
        if (getSupportActionBar() != null) {
            toolbar.setNavigationIcon(getDrawable(R.drawable.baseline_arrow_back_24));
            toolbar.setNavigationOnClickListener(v -> onBackPressed());
        }

        // Set toolbar title
        TextView toolbarTitle = findViewById(R.id.toolbarTitle);
        toolbarTitle.setText("Course Details");

        // Set up profile button
        ImageButton buttonProfile = findViewById(R.id.buttonProfile);
        buttonProfile.setOnClickListener(this::showProfileOptions);

        // Get data from intent
        String courseId = getIntent().getStringExtra("COURSE_ID");
        if (courseId == null) return;

        String courseName = getIntent().getStringExtra("COURSE_NAME");
        if (courseName == null) courseName = "Unknown Course";

        String courseDescription = getIntent().getStringExtra("COURSE_DESCRIPTION");
        if (courseDescription == null) courseDescription = "No description available.";

        // Display course details
        TextView courseNameText = findViewById(R.id.courseName);
        courseNameText.setText(courseName);

        TextView descriptionText = findViewById(R.id.courseDescription);
        descriptionText.setText(courseDescription);

        // Load and display professors associated with the course
        List<ProfessorData> professors = getProfessorsForCourse(courseId);
        Log.d("FilteredProfessors", "Professors for course: " + professors);

        // Set up RecyclerView with filtered professors
        RecyclerView recyclerView = findViewById(R.id.professorsRecyclerView);
        recyclerView.setLayoutManager(new LinearLayoutManager(this));
        recyclerView.setAdapter(new ProfessorAdapter(this, professors));
    }

    private Map<String, ProfessorData> loadAllProfessors() {
        Map<String, ProfessorData> professorMap = new HashMap<>();

        try {
            // Open professors.csv from assets
            BufferedReader reader = new BufferedReader(new InputStreamReader(getAssets().open("professors.csv")));

            // Skip header
            reader.readLine();

            // Read each line and create a map of professors by ID
            String line;
            while ((line = reader.readLine()) != null) {
                String[] tokens = line.split(",");
                if (tokens.length >= 5) {
                    ProfessorData professor = new ProfessorData(
                            tokens[0].trim(),
                            tokens[1].trim(),
                            tokens[2].trim(),
                            tokens[3].trim(),
                            tokens[4].trim()
                    );
                    professorMap.put(professor.getProfessorId(), professor);
                }
            }
            reader.close();
        } catch (Exception e) {
            Log.e("CSVError", "Error loading professors: " + e.getMessage());
        }

        return professorMap;
    }

    private List<ProfessorData> getProfessorsForCourse(String classId) {
        List<ProfessorData> professors = new ArrayList<>();

        try {
            // Load all professors
            Map<String, ProfessorData> allProfessors = loadAllProfessors();

            // Open id.csv to map professors to classes
            BufferedReader readerMappings = new BufferedReader(new InputStreamReader(getAssets().open("id.csv")));

            // Skip header
            readerMappings.readLine();

            // Filter professors associated with the course
            String line;
            while ((line = readerMappings.readLine()) != null) {
                String[] tokens = line.split(",");
                if (tokens.length == 2 && tokens[0].trim().equals(classId)) {
                    ProfessorData professor = allProfessors.get(tokens[1].trim());
                    if (professor != null) {
                        professors.add(professor);
                    }
                }
            }
            readerMappings.close();

        } catch (Exception e) {
            Log.e("CSVError", "Error getting professors for course: " + e.getMessage());
        }

        return professors;
    }

    // Function to display profile options menu
    private void showProfileOptions(View view) {
        PopupMenu popupMenu = new PopupMenu(this, view);
        popupMenu.getMenuInflater().inflate(R.menu.profile_menu, popupMenu.getMenu());

        popupMenu.setOnMenuItemClickListener(item -> {
            if (item.getItemId() == R.id.menu_profile) {
                startActivity(new Intent(CourseDetailActivity.this, ProfileActivity.class));
                return true;
            } else if (item.getItemId() == R.id.menu_logout) {
                Toast.makeText(CourseDetailActivity.this, "Logged out", Toast.LENGTH_SHORT).show();
                return true;
            } else {
                return false;
            }
        });

        popupMenu.show();
    }
}

File: app/src/main/java/com/example/coursecomrade/CourseDetailActivity.kt
package com.example.coursecomrade

import android.content.Intent
import android.os.Bundle
import android.widget.TextView
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.Toolbar
import android.util.Log
import android.view.MenuItem
import android.view.View
import android.widget.ImageButton
import android.widget.PopupMenu
import android.widget.Toast
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import java.io.BufferedReader
import java.io.InputStreamReader
import com.example.coursecomrade.ProfessorAdapter
import com.example.coursecomrade.R

class CourseDetailActivity : AppCompatActivity() {

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_course_detail)

        // Set up toolbar
        val toolbar: Toolbar = findViewById(R.id.toolbar)
        setSupportActionBar(toolbar)
        toolbar.navigationIcon = getDrawable(R.drawable.baseline_arrow_back_24)
        toolbar.setNavigationOnClickListener { onBackPressed() }

        // Set toolbar title
        val toolbarTitle: TextView = findViewById(R.id.toolbarTitle)
        toolbarTitle.text = "Course Details"

        // Set up profile button
        val buttonProfile: ImageButton = findViewById(R.id.buttonProfile)
        buttonProfile.setOnClickListener { showProfileOptions(it) }

        // Get data from intent
        val courseId = intent.getStringExtra("COURSE_ID") ?: return
        val courseName = intent.getStringExtra("COURSE_NAME") ?: "Unknown Course"
        val courseDescription = intent.getStringExtra("COURSE_DESCRIPTION") ?: "No description available."

        // Display course details
        val courseNameText: TextView = findViewById(R.id.courseName)
        courseNameText.text = courseName
        val descriptionText: TextView = findViewById(R.id.courseDescription)
        descriptionText.text = courseDescription

        // Load and display professors associated with the course
        val professors = getProfessorsForCourse(courseId)
        Log.d("FilteredProfessors", "Professors for course: $professors")

        // Set up RecyclerView with filtered professors
        val recyclerView: RecyclerView = findViewById(R.id.professorsRecyclerView)
        recyclerView.layoutManager = LinearLayoutManager(this)
        recyclerView.adapter = ProfessorAdapter(this, professors)
    }

    private fun loadAllProfessors(): Map<String, ProfessorData> {
        val professorMap = mutableMapOf<String, ProfessorData>()

        try {
            // Open professors.csv from assets
            val inputStream = assets.open("professors.csv")
            val reader = BufferedReader(InputStreamReader(inputStream))

            // Skip header
            reader.readLine()

            // Read each line and create a map of professors by ID
            var line: String?
            while (reader.readLine().also { line = it } != null) {
                val tokens = line!!.split(",").map { it.trim() }
                if (tokens.size >= 5) {
                    val professor = ProfessorData(
                        professorId = tokens[0],
                        firstName = tokens[1],
                        lastName = tokens[2],
                        email = tokens[3],
                        rating = tokens[4]
                    )
                    professorMap[professor.professorId] = professor
                }
            }
            reader.close()
        } catch (e: Exception) {
            Log.e("CSVError", "Error loading professors: ${e.message}")
        }

        return professorMap
    }

    private fun getProfessorsForCourse(classId: String): List<ProfessorData> {
        val professors = mutableListOf<ProfessorData>()

        try {
            // Load all professors
            val allProfessors = loadAllProfessors()

            // Open id.csv to map professors to classes
            val inputStreamMappings = assets.open("id.csv")
            val readerMappings = BufferedReader(InputStreamReader(inputStreamMappings))

            // Skip header
            readerMappings.readLine()

            // Filter professors associated with the course
            var line: String?
            while (readerMappings.readLine().also { line = it } != null) {
                val tokens = line!!.split(",").map { it.trim() }
                if (tokens.size == 2 && tokens[0] == classId) {
                    val professor = allProfessors[tokens[1]]
                    if (professor != null) {
                        professors.add(professor)
                    }
                }
            }
            readerMappings.close()

        } catch (e: Exception) {
            Log.e("CSVError", "Error getting professors for course: ${e.message}")
        }

        return professors
    }

    // Function to display profile options menu
    private fun showProfileOptions(view: View) {
        val popupMenu = PopupMenu(this, view)
        popupMenu.menuInflater.inflate(R.menu.profile_menu, popupMenu.menu)

        popupMenu.setOnMenuItemClickListener { item: MenuItem ->
            when (item.itemId) {
                R.id.menu_profile -> {
                    startActivity(Intent(this, ProfileActivity::class.java))
                    true
                }
                R.id.menu_logout -> {
                    Toast.makeText(this, "Logged out", Toast.LENGTH_SHORT).show()
                    true
                }
                else -> false
            }
        }

        popupMenu.show()
    }
}
